Implementation for query function in FindMeetingQuery class

anna-giang opened the pull request:
Given a collection of all known events and a meeting request, `query` will return a collection of meeting times, if any, when the meeting may be scheduled.
All mandatory attendees will be available during returned times. Optional attendees will be included if at least one time slot exists where all mandatory and optional attendees are available, otherwise all optional attendees will be excluded (as per specifications in Week 5 walkthrough.)
Tests that cover the different cases this algorithm needs to handle have been added.

anna-giang committed:
Implementation for query() in FindMeetingQuery class

query() function will find all the time ranges where all attendees are free to have the meeting.
query() functions passes all unit tests in FindMeetingQueryTest, as per first stage of Week 5 TDD Walkthrough.

Algorithm:
1. Going through all the events, add events that involve meeting attendees to an ArrayList, eventList
2. Sort eventList according to event start times
3. Merge adjacent events if they overlap
4. Going through the merged events list, add the gaps between events that are longer or equal to meeting duration to output

anna-giang committed:
Add support for meeting optional attendees

Consider optional attendees in the algorithm to find meeting times.

NOTE: Optional attendees will be ignored completely if their lack of availability causes there to be no meeting times available (in which case only mandatory attendees' schedules will be considered.)

Algorithm is as follows:
1. Going through all the events, add events that involve meeting attendees (both optional and mandatory) to an ArrayList, eventList.
2. Sort eventList according to event start times
3. Merge adjacent events if they overlap
4. Going through the merged events list, add the gaps between events that are longer or  equal to meeting duration to output
5. If there are no meeting times, remove optional attendees, and repeat steps 1-4.

To facilitate the repeat of steps of 1-4, the original query function was also broken up into smaller functions to avoid repeated code.

Also added tests to verify the correctness of the support for optional attendees. All tests pass.

austinct commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,155 @@
+  /**
+   * Returns the subset of events in which the attendees provided are attending.
+   * @param events the collection of Event objects to be filtered
+   * @param attendees the attendees to filter the Event objects by
```
missing an `@return` here

anna-giang replied:
Done! 
I was thinking that the method description already covered what the method will return, which is why I thought maybe an @return statement was not necessary, but I guess it is the best practice for Javadoc.

austinct commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,155 @@
+   * at least of the given duration.
+   * @param eventTimes the sorted list of TimeRanges that represent the time periods in the day
+   *     when events take place
+   * @param duration the minimum length of the period of availability 
```
`@return` here too

anna-giang replied:
Done!

austinct commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,155 @@
+
+    // Finally, add the timeslot from the end of the last event to the end of the day
+    TimeRange lastEvent = eventTimes.get(eventTimes.size() - 1);
+    if (TimeRange.END_OF_DAY - lastEvent.end() >= duration) {
```
Instead of starting from `lastEvent.end()`, you could just start from `timeslotStart` - this would handle the no meetings case properly, so you can remove that edge case at the beginning of this function. Then, this function would also properly handle `duration` greater than an entire day, so you wouldn't need the special case on line 40 too.

anna-giang replied:
Ooh yes, didn't pick that up. Thank you!

anna-giang committed:
Add missing @return statements in Javadocs

anna-giang committed:
Change last timeslot of availability added to be timeslotStart

timeslotStart will be set to the end of the last event OR the start of the day if no events, so the last timeslot of availability will be either from the end of the last event until the end of the day OR  from the start of the day until the end of the day, when there are no events.
This avoids needing to retrieve the last event from the list of events, and also handles the edge case where there are no events.

anna-giang committed:
Update Javadoc of query() function
